### ISSUE TITLE
Test 587 failure

### ISSUE THREAD

**Issue by wizeman** [curl/curl#9007]
### I did this

I was building curl and running the test suite.

### I expected the following

I expected all tests to pass. Instead, I ran into a failure on test 587:

```
test 0581...OK (499 out of 1487, remaining: 05:20, took 0.110s, duration: 02:41)
Use of uninitialized value $USER in concatenation (.) or string at ./runtests.pl line 2224.

Warning: sshserver.pl unknown parameter: 
Will not run ssh server without a user name

Warning: sshserver.pl unknown parameter: 
Will not run ssh server without a user name

Warning: sshserver.pl unknown parameter: 
Will not run ssh server without a user name

Warning: sshserver.pl unknown parameter: 
Will not run ssh server without a user name

Warning: sshserver.pl unknown parameter: 
Will not run ssh server without a user name

Warning: sshserver.pl unknown parameter: 
Will not run ssh server without a user name

Warning: sshserver.pl unknown parameter: 
Will not run ssh server without a user name

Warning: sshserver.pl unknown parameter: 
Will not run ssh server without a user name

Warning: sshserver.pl unknown parameter: 
Will not run ssh server without a user name

Warning: sshserver.pl unknown parameter: 
Will not run ssh server without a user name
RUN: failed to start the SSH server on 22244
RUN: couldn't start SSH. Tried these ports:RUN: 20023, 20243, 20658, 20658, 21142, 21339, 21606, 21707, 21832, 22244Use of uninitialized value $USER in substitution iterator at ./runtests.pl line 3428.
Use of uninitialized value $USER in substitution iterator at ./runtests.pl line 3428.
test 0584...OK (502 out of 1487, remaining: 10:42, took 0.333s, duration: 05:27)
test 0585...OK (503 out of 1487, remaining: 10:41, took 0.245s, duration: 05:27)
test 0586...OK (504 out of 1487, remaining: 10:39, took 0.132s, duration: 05:27)
test 0587...
 587: protocol FAILED!
 There was no content at all in the file log/server.input.
 Server glitch? Total curl failure? Returned: 42
test 0588...OK (506 out of 1487, remaining: 10:36, took 0.153s, duration: 05:28)
```
(...)
```
test 3025...OK (1487 out of 1487, remaining: 00:00, took 0.144s, duration: 12:26)
TESTDONE: 1503 tests were considered during 747 seconds.
TESTDONE: 1193 tests out of 1194 reported OK: 99%

TESTFAIL: These test cases failed: 587 

make[1]: *** [Makefile:810: quiet-test] Error 1
make[1]: Leaving directory '/build/curl-7.83.1/tests'
make: *** [Makefile:1686: test] Error 2
```

### curl/libcurl version

7.83.1

### operating system

Linux hostname 5.15.45 #1-NixOS SMP Mon Jun 6 06:43:42 UTC 2022 x86_64 GNU/Linux

**Comment by bagder** [comment]
It looks like the test script can't find a user name, so it passed on a blank that the script that starts the ssh server doesn't like.

Do you have any of the environment variables `USER`, `USERNAME` or `LOGNAME` set to a user name when the test is  started?

**Comment by wizeman** [comment]
> It looks like the test script can't find a user name, so it passed on a blank that the script that starts the ssh server doesn't like.

Aren't those messages related to tests 582 and 583 (and possibly 584)?

> Do you have any of the environment variables `USER`, `USERNAME` or `LOGNAME` set to a user name when the test is started?

No, NixOS / nixpkgs doesn't set those variables before running the tests.

Note, however, that test 587 usually succeeds. So this appears to be an intermittent (and probably rare) failure, but it would be great to root cause it / fix it, so that it doesn't break builds.

See below for what the test log usually looks like on my system:
```
test 0581...OK (499 out of 1487, remaining: 04:30, took 0.019s, duration: 02:16)
Use of uninitialized value $USER in concatenation (.) or string at ./runtests.pl line 2224.

Warning: sshserver.pl unknown parameter: 
Will not run ssh server without a user name

Warning: sshserver.pl unknown parameter: 
Will not run ssh server without a user name

Warning: sshserver.pl unknown parameter: 
Will not run ssh server without a user name

Warning: sshserver.pl unknown parameter: 
Will not run ssh server without a user name

Warning: sshserver.pl unknown parameter: 
Will not run ssh server without a user name

Warning: sshserver.pl unknown parameter: 
Will not run ssh server without a user name

Warning: sshserver.pl unknown parameter: 
Will not run ssh server without a user name

Warning: sshserver.pl unknown parameter: 
Will not run ssh server without a user name

Warning: sshserver.pl unknown parameter: 
Will not run ssh server without a user name

Warning: sshserver.pl unknown parameter: 
Will not run ssh server without a user name
RUN: failed to start the SSH server on 22244
RUN: couldn't start SSH. Tried these ports:RUN: 20023, 20243, 20658, 20658, 21142, 21339, 21606, 21707, 21832, 22244Use of uninitialized value $USER in substitution iterator at ./runtests.pl line 3428.
Use of uninitialized value $USER in substitution iterator at ./runtests.pl line 3428.
test 0584...OK (502 out of 1487, remaining: 09:52, took 0.200s, duration: 05:01)
test 0585...OK (503 out of 1487, remaining: 09:50, took 0.101s, duration: 05:01)
test 0586...OK (504 out of 1487, remaining: 09:48, took 0.069s, duration: 05:01)
test 0587...OK (505 out of 1487, remaining: 09:47, took 0.096s, duration: 05:02)
test 0588...OK (506 out of 1487, remaining: 09:45, took 0.065s, duration: 05:02)
test 0589...OK (507 out of 1487, remaining: 09:44, took 0.144s, duration: 05:02)
```
(...)
```
test 3019...OK (1481 out of 1487, remaining: 00:02, took 0.013s, duration: 11:43)
test 3020...OK (1482 out of 1487, remaining: 00:02, took 0.013s, duration: 11:43)
Use of uninitialized value $USER in substitution iterator at ./runtests.pl line 3428.
Use of uninitialized value $USER in substitution iterator at ./runtests.pl line 3428.
test 3025...OK (1487 out of 1487, remaining: 00:00, took 0.080s, duration: 11:43)
TESTDONE: 1503 tests were considered during 704 seconds.
TESTDONE: 1194 tests out of 1194 reported OK: 100%
make[1]: Leaving directory '/build/curl-7.83.1/tests'
checkPhase completed in 15 minutes 2 seconds
```

**Comment by bagder** [comment]
I searched for the "Will not run ssh server without a user name" error message and it comes from here:

https://github.com/curl/curl/blob/4572489c5946b5a99d20d6b3692282b07fd4eccd/tests/sshserver.pl#L200

Then I noticed that the `$username` variable it check for is set to the contents of the environment variable `USER` or with the argument `--user`. Then I looked into `runtests.pl` how it invokes the sshserver.pl script and there I saw:

https://github.com/curl/curl/blob/4572489c5946b5a99d20d6b3692282b07fd4eccd/tests/runtests.pl#L2228

So it *does* pass in `--user "$USER"` to the script. So what is $USER set to?

https://github.com/curl/curl/blob/4572489c5946b5a99d20d6b3692282b07fd4eccd/tests/runtests.pl#L389-L396

Meaning: if it cannot find a user name, it will fail to start the ssh server.

A fix should probably prevent it from even trying if there's no user name, rework how it figures out the user name or if possible perhaps fix the sshserver.pl script to work without a user name.
